src/rapid_pro_to_engagement_db/rapid_pro_to_engagement_db.py

```diff
@@ -0,0 +1,155 @@
+from core_data_modules.cleaners import URNCleaner
+from core_data_modules.logging import Logger
+from engagement_database.data_models import Message, MessageDirections, MessageStatuses, HistoryEntryOrigin
+from rapid_pro_tools.rapid_pro_client import RapidProClient
+from storage.google_cloud import google_cloud_utils
+
+log = Logger(__name__)
+
+
+def init_rapid_pro_client(google_cloud_credentials_file_path, rapid_pro_domain, rapid_pro_token_file_url):
+    """
+    Initialises a RapidProClient from a Rapid Pro domain and token file url.
+
+    :param google_cloud_credentials_file_path: Path to the google cloud credentials file to use to access the
+                                               rapid pro token.
+    :type google_cloud_credentials_file_path: str
+    :param rapid_pro_domain: Domain which Rapid Pro is running on.
+    :type rapid_pro_domain: str
+    :param rapid_pro_token_file_url: GS url to the Rapid Pro token.
+    :type rapid_pro_token_file_url: str
+    :return: RapidProClient for
+    :rtype: rapid_pro_tools.rapid_pro_client.RapidProClient
+    """
+    log.info(f"Initialising Rapid Pro client for domain {rapid_pro_domain} and auth "
+             f"url {rapid_pro_token_file_url}...")
+    rapid_pro_token = google_cloud_utils.download_blob_to_string(
+        google_cloud_credentials_file_path, rapid_pro_token_file_url).strip()
+    rapid_pro_client = RapidProClient(rapid_pro_domain, rapid_pro_token)
+    log.info("Initialised Rapid Pro client")
+
+    return rapid_pro_client
+
+
+def _engagement_db_has_message(engagement_db, message):
+    """
+    Checks if an engagement database contains a message with the same text, timestamp, and participant_uuid as the
+    given message.
+
+    :param engagement_db: Engagement database to check for the message.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param message: Message to check for existence.
+    :type message: engagement_database.data_models.Message
+    :return: Whether a message with this text, timestamp, and participant_uuid exists in the engagement database.
+    :rtype: bool
+    """
+    # TODO: Rapid Pro has a bug where timestamps occasionally drift by 1us when runs are archived.
+    #       Confirm this is resolved before entering production with an '==' check on timestamps.
+    filter = lambda q: q \
```

Comment:
Checking for exists for single messages will be expensive as we scale. We should probably fetch message ids in cache and use this for the check.

Reply:
I don't think this cost is very high, as in incremental mode we're only paying 1 read per incoming message by doing this.  How would a cache help?

Reply:
I was thinking  - For the Initial rapid pro sync, we store message ids for a project in the cache, for the next subsequent run we check if the new message-id exists in the cached Ids and update the list write to firestore if it does not. This means we do a full read from a project firestore once at the pipeline initial run but read and write only when we have new messages.

Reply:
resolved offline